File: typescript/lesson-4.ts
// Lesson 4: Objects

// adding export to be used in other module
type Payment = { // object should be UPPERCASE, use type (instead of const)
    referenceId: string // no need comma on this
    amount: number
    paymentMethod: string
    isSettled: boolean
}

const txn1: Payment = {
    referenceId: "txn1", //this is like JSON where we need "," on each param
    amount: 10000,
    paymentMethod: "Cards",
    isSettled: true
}

const txn2: Payment = {
    referenceId: "txn2",
    amount: 50000,
    paymentMethod: "DANA",
    isSettled: false
}


// Create function
function describePayment (
    payment: Payment
): string {
    return "Payment " + payment.referenceId + " via " + payment.paymentMethod + " with amount of " + payment.amount + " is " + (payment.isSettled === true ? "settled" : "not settled") //condition ? "value if true" : "value if false"
}

// print
console.log(describePayment(txn1))
console.log(describePayment(txn2))

File: golang/lesson-4.go
package main

import "fmt"

// create the type
type Payment struct {
	ReferenceId   string
	Amount        int
	PaymentMethod string
	IsSettled     bool
}

// create the vars, 2 vars
var pymts1 = Payment{
	ReferenceId:   "txn01",
	Amount:        50000,
	PaymentMethod: "OVO",
	IsSettled:     true,
}

var pymts2 = Payment{
	ReferenceId:   "txn02",
	Amount:        10000,
	PaymentMethod: "Cards",
	IsSettled:     false,
}

func describePayment(p Payment) string { // use the one word only to simplified

	var status string
	if p.IsSettled {
		status = "Settled"
	} else {
		status = "Unsettled"
	}

	// note on above, i can use status := "Unsettled" then put the if condition, but only if i declare the value first, not an empty information

	return "Payment " + p.ReferenceId +
		" from " + p.PaymentMethod +
		" with amount " + fmt.Sprintf("%d", p.Amount) +
		" is " + status // if else simplified
}

// call the func
func main() {
	fmt.Println(describePayment(pymts1))
	fmt.Println(describePayment(pymts2))
}
